app/controllers/application_controller.rb

```diff
@@ -1,10 +1,157 @@
+    else
+      redirect "/tweets"
+    end
+  end
+
+  post '/signup' do
+    user = User.new(username: params["username"], email: params["email"], password: params["password"])
+    if !user.username.empty? && !user.email.empty? && user.save
+      session[:id] = user.id
+      redirect to "/tweets"
+    elsif params["username"].empty? || params["email"].empty?
+      redirect to "/signup", :locals => {:missing => "Please fill in all fields"}
+    else
+      redirect to "/signup", :locals => {:error => "Not a valid username/password. Please try again"}
+    end
+  end
+
+  get '/login' do 
+    if !logged_in?
+      erb :"users/login"
+    else
+      redirect to '/tweets'
+    end
+  end
+
+  post '/login' do
+    if !params["username"].empty?
+      user = User.find_by(username: params["username"])
+      if !!user && user.authenticate(params["password"])
+        session[:id] = user.id  
+        redirect to "/tweets"
+      end
+    else
+      redirect to "/login"
+    end
+  end
+
+  get "/tweets" do
+    if logged_in?
+      @tweets = Tweet.all
+      erb :"tweets/index"
+    else
+      redirect to '/login'
+    end
+  end
+
+  get "/logout" do 
+    session.clear
+    redirect to '/login'
+  end
+
+  get "/users/:slug" do
+    erb :"users/show" if params[:slug] == current_user.slug
+  end
+
+  get "/tweets/new" do 
+    if logged_in?
+      erb :"tweets/new"
+    else
+      redirect to "/login"
+    end
+  end
+
+  post "/tweets" do
+    if logged_in?
+      if !params["content"].empty?
+        @new_tweet = Tweet.create(content: params["content"], user_id: session[:id])
+        redirect to "/tweets"
+      else
+        redirect to "/tweets/new"
+      end
+    else
+      redirect to "/login"
+    end
+  end
+
+  get "/tweets/:id" do
+    if logged_in?
+      @user = current_user
+      @tweet = Tweet.find(params[:id])
+      erb :"tweets/show"
+    else
+      redirect to "/login"
+    end
+  end
+
+  get "/tweets/:id/edit" do
+    if logged_in?
+      @user = current_user
+      @tweet = Tweet.find(params[:id])
+      if @user.id == @tweet.user_id
+        erb :"tweets/edit"
+      else
+        redirect to "/tweets/#{params[:id]}"
+      end
+    else
+      redirect to "/login"
+    end
+  end
+
+  post "/tweets/:id" do
+    if logged_in?
+      if !params["content"].empty?
+        @tweet = Tweet.find(params[:id])
+        @tweet.update(content: params["content"])
+      else
+        redirect to "/tweets/#{params[:id]}/edit"
+      end
+    else
+      redirect to "/login"
+    end
+  end
+
+  post "/tweets/:id/delete" do
+    if logged_in?
+      @tweet= Tweet.find(params[:id])
+      if @tweet.user_id == session[:id]
+        @tweet.destroy
+        redirect to "/tweets"
+      else
+        redirect to "/tweets"
+      end
+    else
+      redirect to "/login"
+    end
+  end
+
+  helpers do
+    def logged_in?
+      !!session[:id]
+    end
+
+    def current_user
+      User.find(session[:id])
```

Comment:
Because in the lifespan of a request, you might call `current_user` multiple times, we'd prefer this method to only hit the database once and then 'cache' the user instance into an instance variable that could be returned on any subsequent call to `current_user`.

```ruby
def current_user
  @current_user ||= User.find(session[:id) if session[:id]
end
```

This pattern of conditionally setting an instance variable or returning it if it's already defined using the `||=` (or-equals) operator is called Memoization. ([read more](http://gavinmiller.io/2013/basics-of-ruby-memoization/) and [here](http://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/))